src/video/video.c

```diff
@@ -1,306 +1,174 @@
-
-    // Activate RGBA conversion program, and bind palette etc.
-    activate_program(g_video_state.rgba_prog_id);
-    bind_uniform_4fv(g_video_state.rgba_prog_id, "projection", projection_matrix);
-    GLuint pal_ubo_id = shared_get_block(g_video_state.shared);
-    bind_uniform_block(g_video_state.rgba_prog_id, "palette", PAL_BLOCK_BINDING, pal_ubo_id);
-    bind_uniform_1i(g_video_state.rgba_prog_id, "framebuffer", TEX_UNIT_FBO);
-    bind_uniform_1i(g_video_state.rgba_prog_id, "remaps", TEX_UNIT_REMAPS);
-
-    INFO("OpenGL Renderer initialized!");
-    return 0;
-
-error_3:
-    delete_program(g_video_state.palette_prog_id);
-
-error_2:
-    SDL_GL_DeleteContext(g_video_state.gl_context);
-
-error_1:
-    SDL_DestroyWindow(g_video_state.window);
-
-error_0:
-    return 1;
 }
 
-void video_draw_atlas(bool draw_atlas) {
-    g_video_state.draw_atlas = draw_atlas;
+bool video_get_renderer_info(int index, bool *is_available, const char **name, const char **description) {
+    if(index < 0 && index >= renderer_count)
+        return false;
+    if(is_available != NULL)
+        *is_available = renderer_statuses[index].is_available;
+    if(name != NULL)
+        *name = renderer_statuses[index].name;
+    if(description != NULL)
+        *description = renderer_statuses[index].description;
+    return true;
 }
 
-void video_reinit_renderer(void) {
+int video_get_renderer_count(void) {
+    return renderer_count;
 }
 
-int video_reinit(int window_w, int window_h, bool fullscreen, bool vsync) {
-    g_video_state.screen_w = window_w;
-    g_video_state.screen_h = window_h;
-    g_video_state.fullscreen = fullscreen;
-    g_video_state.vsync = vsync;
-    bool success = resize_window(g_video_state.window, window_w, window_h, fullscreen);
-    success = set_vsync(g_video_state.vsync) && success;
-
-    // Fetch viewport size which may be different from window size.
-    SDL_GL_GetDrawableSize(g_video_state.window, &g_video_state.viewport_w, &g_video_state.viewport_h);
-
-    return success;
+static bool hunt_renderer_by_name(const char *try_name) {
+    for(int i = 0; i < renderer_count; i++) {
+        if(strcmp(renderer_statuses[i].name, try_name) != 0)
+            continue;
+        if(!renderer_statuses[i].is_available)
+            continue;
+        renderer_statuses[i].init(&current_renderer);
+        return true;
+    }
+    return false;
 }
 
-// Called on every game tick
-void video_reset_atlas(void) {
-    atlas_reset(g_video_state.atlas);
+static bool find_best_renderer(void) {
+    for(int i = 0; i < renderer_count; i++) {
+        if(renderer_statuses[i].is_available) {
+            renderer_statuses[i].init(&current_renderer);
+            return true;
+        }
+    }
+    return false;
 }
 
-void video_render_prepare(void) {
-    object_array_prepare(g_video_state.objects);
+static bool video_find_renderer(const char *try_name) {
+    memset(&current_renderer, 0, sizeof(renderer));
```

Comment:
if video_find_renderer is to be called at most once per program run, it does not need to zero current_renderer-- C guarantees static objects are initialized.
If video_find_renderer can be called multiple times per program run (e.g. switching renderers in the settings menu), it should close the previous current_renderer before memsetting it to zero to prevent a leak.

same applies to find_best_renderer.

Reply:
None of these functions do any memory allocations; they just swap function pointers in the current_renderer struct.

But yeah, the memsets were debugging stuff; I will get rid of them.

Reply:
Reorganized this part to be clearer.